File: designpatternsfromjude/src/com/renho/structural/decorator/Decorator.java
package com.renho.structural.decorator;

public class Decorator extends Component {

	private Component component;

	public Decorator() {}
	
	public Decorator(Component component) {
		this.component = component;
	}

	public void operation() {
		this.component.operation();
	}

	public Component getComponent() {
		return component;
	}

	public void setComponent(Component component) {
		this.component = component;
	}
}

File: ideaworkspace/designpatterns/src/main/scala/com/renho/structural/decorator/Decorator.scala
package com.renho.structural.decorator

/**
  * Created by renho on 2016/7/11.
  */
class Decorator(component: Component) extends Component {
	override def operation(): Unit = {
		this.component.operation()
	}
}

class ConcreteDecoratorA(component: Component) extends Decorator(component: Component) {
	def addBehavior() {
		println("ConcreteDecoratorA addBehavior");
	}
	override def operation(): Unit = {
		addBehavior()
		this.component.operation()
	}
}

class ConcreteDecoratorB(component: Component) extends Decorator(component: Component) {
	def addBehavior() {
		println("ConcreteDecoratorB addBehavior");
	}
	override def operation(): Unit = {
		addBehavior()
		this.component.operation()
	}
}
